dashboard/app/controllers/omniauth_callbacks_controller.rb

```diff
@@ -174,14 +174,29 @@ def just_authorized_google_classroom(user, params)
       scopes.include?('classroom.rosters.readonly')
   end
 
-  def silent_takeover(oauth_user)
+  def silent_takeover(oauth_user, auth_hash)
     # Copy oauth details to primary account
     @user = User.find_by_email_or_hashed_email(oauth_user.email)
-    @user.provider = oauth_user.provider
-    @user.uid = oauth_user.uid
-    @user.oauth_refresh_token = oauth_user.oauth_refresh_token
-    @user.oauth_token = oauth_user.oauth_token
-    @user.oauth_token_expiration = oauth_user.oauth_token_expiration
+    if @user.migrated?
+      AuthenticationOption.new(
+        user: @user,
+        email: oauth_user.email,
+        hashed_email: oauth_user.hashed_email,
+        credential_type: auth_hash.provider.to_s,
+        authentication_id: auth_hash.uid,
+        data: {
+          oauth_token: auth_hash.credentials&.token,
+          oauth_token_expiration: auth_hash.credentials&.expires_at,
+          oauth_refresh_token: auth_hash.credentials&.refresh_token
+        }
+      ).save!
+    else
+      @user.provider = oauth_user.provider
+      @user.uid = oauth_user.uid
+      @user.oauth_refresh_token = oauth_user.oauth_refresh_token
+      @user.oauth_token = oauth_user.oauth_token
+      @user.oauth_token_expiration = oauth_user.oauth_token_expiration
+    end
```

Comment:
Ah, that's right, I remember you mentioned that. In theory, https://github.com/code-dot-org/code-dot-org/blob/f74ef81f1fd8584d8225be8375bbb7be9c45ab93/dashboard/app/controllers/omniauth_callbacks_controller.rb#L78 should be where we end up going if the user already exists and the lookup succeeds (`user.persisted?` should be true), in which case we call `check_and_apply_oauth_takeover`, which is different from the silent takeover logic below it. I'll verify this, though, and maybe write some tests to check that logic.

Reply:
Oooooh - and `check_and_apply_oauth_takeover` is actually a  no-op unless we have specific session variables in place. So that's the plain 'sign in' case - it just doesn't read like that yet.  TIL.

Reply:
Yeah, this could really use some refactoring to make it clear what each part does.

Reply:
Note for the future: we discussed this offline, and decided that hoisting the if condition in `check_and_apply_oauth_takeover` up to `login` would make this quite a bit more clear. TODO in another PR.